### PR TITLE
colexec: improve sort operator a bit

### PR DESCRIPTION
**Pull request by yuzefovich**
**colexec: improve sort operator a bit**

Previously, we were missing memory accounting around some slices
allocated internally by the sort operator which is now added.
Additionally, the references to those slices are now kept by the
operator which will be useful when the sorter is used by the external
sorter.

Additionally, this commit eliminates some bounds checks.

Release note: None

**colexec: add some BCE assertions for distinct and sort**

This is achieved by templating the inlined execgen functions. Note that
in some cases this templating leaves redundant `_ = true` lines, but
figuring out what's up with that is left as a TODO.

Additionally, this commit removes some redundant attempts at achieving
BCE for non-sliceable vectors (which had a negative impact in case of
JSONs) and clarifies the name of a utility function.

Release note: None

### CODE REVIEW

**Comment by DrewKimball** [comment]
@yuzefovich where are the assertion comments omitted? I probably have enough context to fix it fairly easily.

**Comment by yuzefovich** [comment]
I think the best way to see them is through the web browser on github.com, if you look at `Files changed` and click `Load diff` for `distinct.eg.go`.

**Comment by yuzefovich** [comment]
As Drew correctly pointed out, the gcassert comments are always added into the generated code, but sometimes there are useless `_ = true` lines. I think it's ok to leave them as is since they should be compiled away. I also ran the microbenchmarks for the external sort, and they didn't show much of change (we're only using int64 though).

File: pkg/sql/colexec/colexecbase/distinct_tmpl.go [hunk 1]
```diff
@@ -259,19 +260,23 @@ func (p partitioner_TYPE) partitionWithOrder(
...
 	_ = outputCol[n-1]
-	// TODO(yuzefovich): add BCE assertions for these.
+	_ = order[n-1]
 	outputCol[0] = true
 	if nulls != nil {
 		for outputIdx := 0; outputIdx < n; outputIdx++ {
+			//gcassert:bce
 			checkIdx := order[outputIdx]
-			lastVal, lastValNull = checkDistinctWithNulls(checkIdx, outputIdx, lastVal, nulls, lastValNull, col, outputCol, p.nullsAreDistinct)
+			lastVal, lastValNull = checkDistinctWithNulls(checkIdx, outputIdx, lastVal, nulls, lastValNull, col, outputCol, p.nullsAreDistinct, false, true)
```
Comment (DrewKimball) [inline pkg/sql/colexec/colexecbase/distinct_tmpl.go:273]: I think we're missing a BCE here - `colBCE` should be true.

File: pkg/sql/colexec/colexecbase/distinct_tmpl.go [hunk 2]
```diff
@@ -259,19 +260,23 @@ func (p partitioner_TYPE) partitionWithOrder(
...
 		}
 	} else {
 		for outputIdx := 0; outputIdx < n; outputIdx++ {
+			//gcassert:bce
 			checkIdx := order[outputIdx]
-			lastVal = checkDistinct(checkIdx, outputIdx, lastVal, col, outputCol)
+			lastVal = checkDistinct(checkIdx, outputIdx, lastVal, col, outputCol, false, true)
```
Comment (DrewKimball) [inline pkg/sql/colexec/colexecbase/distinct_tmpl.go:279]: ditto

**Review by DrewKimball — APPROVED** [review]
After my two comments are resolved, LGTM.

**Review by mgartner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67638)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/colexec/sort.go, line 419 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67638#-MefJdSS90nT6LLJx8Y0:-MefJdSS90nT6LLJx8Y1:bp1sc0f) ([raw file](https://github.com/cockroachdb/cockroach/blob/55984c360305616f71f0f3977ffda7bb5e47a5bf/pkg/sql/colexec/sort.go#L419)):*
> ```Go
> 		// that were true in the distinct vector.
> 		sizeBefore := memsize.Int * int64(cap(p.scratch.partitions))
> 		p.scratch.partitions = boolVecToSel64(partitionsCol, p.scratch.partitions[:0])
> ```

Not related to your change, but why is it called `boolVecToSel64` if `[]int` is returned. I'd expect `[]int64` to be returned.

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67638)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @mgartner)

___
*[pkg/sql/colexec/sort.go, line 419 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67638#-MefJdSS90nT6LLJx8Y0:-MefaVO8CUOj7Unc2kDI:byna8ek) ([raw file](https://github.com/cockroachdb/cockroach/blob/55984c360305616f71f0f3977ffda7bb5e47a5bf/pkg/sql/colexec/sort.go#L419)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Not related to your change, but why is it called `boolVecToSel64` if `[]int` is returned. I'd expect `[]int64` to be returned.
</blockquote></details>

The selection vector used to be `[]int64` but it was refactored a couple of releases ago. Fixed.
___
*[pkg/sql/colexec/colexecbase/distinct_tmpl.go, line 273 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67638#-MebU-N39W1Mc7oUftr--r2-273:-MefbOb6B-VcYaTjxM2K:b-jczy26) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b8501541d22e9982fd1e177ae7bed26bd4878f7/pkg/sql/colexec/colexecbase/distinct_tmpl.go#L273)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

I think we're missing a BCE here - `colBCE` should be true.
</blockquote></details>

No, I don't think so - actually, we're accessing `col` with `checkIdx == order[outputIdx]`, and the compiler cannot prove that values in `order` slice are in bounds. Removed the redundant attempt at BCE for it.
___
*[pkg/sql/colexec/colexecbase/distinct_tmpl.go, line 279 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67638#-MebU-N39W1Mc7oUftr--r2-279:-Mefb_Uj9CGwbpENZgHo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b8501541d22e9982fd1e177ae7bed26bd4878f7/pkg/sql/colexec/colexecbase/distinct_tmpl.go#L279)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.

**Comment by yuzefovich** [comment]
TFTRs!

bors r+

**Pull request merged**